In [12]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold

import xgboost as xgb
from sklearn.metrics import average_precision_score as APS
from sklearn.ensemble import RandomForestClassifier as SklearnRF

In [38]:
FILE = '../data/arc_final_des.csv'
# FILE = '../data/arc_final_fin.csv'
# FILE = '../data/arc_final_ms_spec.csv'

data = pd.read_csv(FILE, index_col=0)
data

Floral  Fruity  Herbal  Green  Woody  Sweet  Balsamic  Earth  \
107             0       0       0      0      0      1         1      0   
126             0       0       0      0      1      0         1      0   
174             0       0       0      0      0      0         0      1   
177             0       0       0      0      0      1         0      0   
179             0       0       0      0      0      1         0      0   
...           ...     ...     ...    ...    ...    ...       ...    ...   
140998267       1       0       1      0      0      1         0      0   
141266555       0       1       1      0      0      0         0      0   
142523980       1       0       0      0      1      0         1      0   
144116082       1       0       0      1      1      0         0      0   
145793235       1       1       0      0      0      0         0      0   

           Spicy  Chemical  ...       AMW    WTPT-1    WTPT-2    WTPT-3  \
107            0         0  ...  0.870601 -0.485781  0.428391 -0.097812   
126            0         0  ...  1.750794 -1.013401  0.300976 -0.033139   
174            0         0  ...  0.032445 -2.425877 -3.359749 -0.152376   
177            0         0  ...  0.108736 -2.684755 -4.617587 -0.998939   
179            0         0  ...  0.108736 -1.942763 -2.790262 -0.142069   
...          ...       ...  ...       ...       ...       ...       ...   
140998267      0         0  ...  0.052427  0.621329 -0.971686  0.947798   
141266555      0         0  ... -0.872139  0.654758 -0.722589  0.139547   
142523980      0         0  ... -1.057710  0.027658  0.386557 -0.922173   
144116082      0         0  ... -0.893362  0.236738 -0.035581 -0.909362   
145793235      0         0  ... -0.737443  0.409447 -0.690519 -0.929301   

             WTPT-4    WTPT-5     WPATH      WPOL     XLogP    Zagreb  
107        0.001562 -0.242563 -0.427532 -0.596113 -0.024422 -0.443478  
126        0.067627 -0.242563 -0.633394 -0.737723 -1.165779 -0.836798  
174       -0.054176 -0.242563 -0.829452 -1.870605 -2.584962 -2.311751  
177       -0.918961 -0.242563 -0.844156 -2.012215 -1.772042 -2.508411  
179       -0.043648 -0.242563 -0.782888 -1.445774 -2.119984 -1.721770  
...             ...       ...       ...       ...       ...       ...  
140998267  1.069679 -0.242563  0.344448  0.678379 -0.439317  0.638154  
141266555  0.244031 -0.242563  0.410618  0.395159  0.916077  0.441493  
142523980 -0.840543 -0.242563 -0.167755  0.111938  0.624544  0.048173  
144116082 -0.827456 -0.242563 -0.104036  0.961600  0.232845  0.539824  
145793235 -0.847824 -0.242563  0.317490 -0.029672  0.588168  0.146503  

[2328 rows x 1132 columns]

In [39]:
NUM_LABEL_COLS = 12
FEATURE_COLS = data.columns[NUM_LABEL_COLS:]
LABEL_COLS = data.columns[:NUM_LABEL_COLS]
print(FEATURE_COLS)
print(LABEL_COLS)

X = data[FEATURE_COLS].values
y = data[LABEL_COLS].values

X.shape, y.shape

Index(['nAcid', 'ALogP', 'ALogp2', 'AMR', 'apol', 'naAromAtom', 'nAromBond',
       'nAtom', 'nHeavyAtom', 'nH',
       ...
       'AMW', 'WTPT-1', 'WTPT-2', 'WTPT-3', 'WTPT-4', 'WTPT-5', 'WPATH',
       'WPOL', 'XLogP', 'Zagreb'],
      dtype='object', length=1120)
Index(['Floral', 'Fruity', 'Herbal', 'Green', 'Woody', 'Sweet', 'Balsamic',
       'Earth', 'Spicy', 'Chemical', 'Animalic', 'Nutty'],
      dtype='object')


((2328, 1120), (2328, 12))

In [29]:
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.2, random_state=42)
train_X.shape, test_X.shape, train_y.shape, test_y.shape

((836, 344), (209, 344), (836, 12), (209, 12))

In [31]:
classifier = xgb.XGBRegressor(tree_method='hist', seed=42)
# classifier = SklearnRF(random_state=42)

# Fitting the model
classifier.fit(train_X, train_y)

# Predict the model
pred = classifier.predict(test_X)

aps = APS(test_y, pred)

print("APS : % f" % aps)

APS :  0.331917


<h3>Descriptors</h3>

- XGB: 0.437068
- RF: 0.298366


<h3>Fingerprints</h3>

- XGB: 0.430697
- RF: 0.295155


<h3>Mass Spectra</h3>

- XGB: 0.331917
- RF: 0.218172
